crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2603,6 +2604,49 @@ impl<'db> TypeInferenceBuilder<'db> {
                     op,
                     Type::IntLiteral(i64::from(b)),
                 ),
+
+            (Type::StringLiteral(s1), Type::StringLiteral(s2)) => match op {
+                ast::CmpOp::Eq => Some(Type::BooleanLiteral(s1 == s2)),
+                ast::CmpOp::NotEq => Some(Type::BooleanLiteral(s1 != s2)),
+                ast::CmpOp::Lt => Some(Type::BooleanLiteral(s1 < s2)),
+                ast::CmpOp::LtE => Some(Type::BooleanLiteral(s1 <= s2)),
+                ast::CmpOp::Gt => Some(Type::BooleanLiteral(s1 > s2)),
+                ast::CmpOp::GtE => Some(Type::BooleanLiteral(s1 >= s2)),
```

Comment:
I don't think this is doing what you intend, since `s1` and `s2` are `StringLiteralType` salsa interned structs. So what you're actually comparing here is just the Salsa u32 IDs. I suspect a few more string comparison tests (specifically, one where the string created earlier in the source file is "greater" than one created later in the source file) will reveal the problem. You'll need to use `s1.value(db)` and `s2.value(db)` like you do below in the "in" branch to actually compare the strings.

Let's not even worry about whether Rust and Python have any weird edge-case differences in their lexicographic string ordering 😆 Probably not, and if so I suspect nobody will ever care anyway.